### PR TITLE
Support specifying tree architectures

### PR DESCRIPTION
**Pull request by Toaster192**
example database index:
```
trees:
  foo:
    template: trees/foo.xml.j2
    arches: .*
  bar:
    template: trees/bar.xml.j2
    arches: .*
  baz:
    template: trees/baz.j2
    arches:
      - x86_64
      - aarch64
      - ppc64.*
```
And error when a case is expection a different arch than the tree has available
```
kpet.schema.Invalid: One of case patterns
match an unavailable tree/arch combination
In suite: /acpi/acpitable test
In case: ACPI table test
The arch: s390x
The tree: baz
The avaliable arches: ['x86_64', 'aarch64', 'ppc64', 'ppc64le']
```
Posting this as a WIP since there is much to discuss and fix at the moment, for example the error above might be unwanted since the test matched all arches and many trees yet this says it's not correct because a single tree doesn't support all arches

I suppose that should instead work so that the exception is raised if the only matching arch is the unavailible one

Also line 410 question: Do we want the 'trees: arches:' regexes to all match a arch or have them combined match at least a single arch?

### CODE REVIEW

File: kpet/data.py [hunk 1]
```diff
@@ -387,6 +388,58 @@ def validate_host_type_regex(self):
...
 
+    def validate_trees_arches(self):
+        """
+        Check if any of the regexes are invalid.
+        Returns:
+            Raises a schema.Invalid exception when finding an invalid regex
+        """
+
+        trees = {}
+        # Create a dict {tree: arches,} where arches is a list or regexes
```
Comment (Toaster192) [inline kpet/data.py]: list of regexes*

File: kpet/cmd_arch.py [hunk 2]
```diff
@@ -30,6 +30,10 @@ def build(cmds_parser, common_parser):
         help='Output a list of known architecture names',
         parents=[common_parser],
     )
+    list_parser.add_argument('-t', metavar='TREE_REGEX',
```
Comment (spbnick) [inline kpet/cmd_arch.py]: Could you make it accept `--tree` as well, just as the `kpet run generate` command does? Also, please don't repeat "TREE" in the "metavar" name. Just "REGEX" would be enough. We already have "tree" in the option name.

File: kpet/cmd_arch.py [hunk 3]
```diff
@@ -30,6 +30,10 @@ def build(cmds_parser, common_parser):
...
+                             default=None,
+                             help='Optional. Use this to filter results '
+                             'by available trees')
```
Comment (spbnick) [inline kpet/cmd_arch.py]: All "options" are supposed to be optional by default, unless specified otherwise, so please don't write that. Also please describe what the option means/does instead of what to use it for. E.g. write: "Limit output to architectures with trees matching the REGEX".

File: kpet/cmd_arch.py [hunk 4]
```diff
@@ -30,6 +30,10 @@ def build(cmds_parser, common_parser):
...
     list_parser.add_argument('regex', nargs='?', default=None,
                              help='Optional. Use this to filter results.')
```
Comment (spbnick) [inline kpet/cmd_arch.py]: Would be good to improve this too.

File: kpet/cmd_arch.py [hunk 5]
```diff
@@ -41,9 +45,20 @@ def main(args):
...
+                    if tree_regex.fullmatch(tree):
+                        matching_arches.append(arch)
+            matching_arches = list(dict.fromkeys(matching_arches))
+        else:
+            matching_arches = database.arches
+
+        for arch in sorted(matching_arches):
             # filter using regex; if it's not set, print all
-            if not regex or regex.match(arch):
+            if not regex or regex.fullmatch(arch):
```
Comment (spbnick) [inline kpet/cmd_arch.py]: Please make a separate commit fixing this.

File: kpet/cmd_tree.py [hunk 6]
```diff
@@ -30,6 +30,10 @@ def build(cmds_parser, common_parser):
         help='List available kernel trees.',
         parents=[common_parser],
     )
+    list_parser.add_argument('-a', metavar='ARCH_REGEX',
+                             default=None,
+                             help='Optional. Use this to filter results '
+                             'by available arches')
```
Comment (spbnick) [inline kpet/cmd_tree.py]: Comments from `cmd_arch.py` apply here too.

File: kpet/data.py [hunk 7]
```diff
@@ -387,6 +388,62 @@ def validate_host_type_regex(self):
                                   host_type_regex.pattern,
                                   host_types)
 
+    # pylint: disable=too-many-branches
+    def validate_trees_arches(self):
```
Comment (spbnick) [inline kpet/data.py]: Functions called "validate" are not expected to return transformed data, but pass or fail instead (even if with an exception). In this case I would rather call it something like "load" and mention that it raises an exception if the data is invalid. However, I think we shouldn't be doing this at all. Please see the comment below.

File: kpet/data.py [hunk 8]
```diff
@@ -400,7 +457,10 @@ def __init__(self, dir_path):
                     ),
                     optional=dict(
                         suites=List(YAMLFile(Class(Suite))),
-                        trees=Dict(String()),
+                        trees=Dict(Dict(Reduction(String(),
+                                                  lambda x: [x],
+                                                  List(String())
+                                                  ))),
```
Comment (spbnick) [inline kpet/data.py]: This is under-specified. Please specify the expected structure strictly. I.e.:

```python
trees=Dict(
    StrictStruct(template=String(),
                 arches=Reduction(Regex(),
                                  lambda x: [x], 
                                  List(Regex()))
    )
)   
```

Or maybe better:

```python
trees=Dict(
    Struct(required=dict(template=String()),
           optional=dict(arches=Reduction(Regex(),
                                          lambda x: [x],
                                          List(Regex())))
    )                                 
)                                     
``` 

And make the missing `arches` default to a list with a single match-all regex. That way we won't have to specify "arches" for every tree, just for the ones which have limited architecture support. 

Put some of the intermediate arguments into variables to reduce nesting depth, if needed.

Then make sure kpet would also read the data in the existing format present in kpet-db, so we can update kpet first and then kpet-db. That could be done using the `Succession` schema. Check one of the recent commits by @wizardofoz123 for an example.

File: kpet/data.py [hunk 9]
```diff
@@ -426,3 +486,11 @@ def __init__(self, dir_path):
...
         # Regex check
         self.validate_host_type_regex()
+        self.tree_arches = self.validate_trees_arches()
+        # Convert self.trees back to just {tree: template,} because we don't
+        # need the arches after validating them
+        for tree in self.trees:
+            if len(self.trees[tree]["template"]) == 1:
+                self.trees[tree] = str(self.trees[tree]["template"][0])
+            else:
+                raise Invalid("Multiple tree templates")
```
Comment (spbnick) [inline kpet/data.py]: Let's not do any conversion here. I think the DB structure is good enough to use, and if we keep the structure the same the code will be easier to follow and reason about.

If necessary, we could maybe define a `Tree` class and encapsulate some methods for matching the tree to an architecture, or maaybe resolve the architecture regex list into an architecture name list, but let's see if we can keep it straightforward first.

**Review by spbnick — CHANGES_REQUESTED** [review]
Thank you, Ondrej, this is a good start :)
I left some comments/requests inline.

I think requiring tests to match only available tree/architecture combinations would make writing the conditions more difficult with little benefit, so let's wait with that.
I don't think we need to implement checking if the arch regexes match anything for the start, but if/when we do it, I think it'll be good to check that every regex in the list matches at least something.

I'm also thinking if it will be enough to only support exact architecture name specification for trees, instead of regexes. There aren't that many of those architectures for us to really need it, and the resulting data might be easier to read. I.e. you won't have to parse the regexes and match them to the architecture list mentally, when reading the `index.yaml`. The code will also be much simpler. Could you ask somebody, or dig it out, what are our actual supported architectures for all the trees (we'll need to do that anyway), and try writing the actual data we need using both methods?

Also, we're missing a check in cmd_run to see if the supplied tree/arch combination is valid.

Thanks!

File: kpet/data.py [hunk 10]
```diff
@@ -387,26 +388,105 @@ def validate_host_type_regex(self):
...
-                        host_types=Dict(Class(HostType)),
-                        host_type_regex=Regex(),
-                        recipesets=Dict(List(String()))
+                Succession(
+                    Struct(
+                        required=dict(
+                        ),
+                        optional=dict(
+                            suites=List(YAMLFile(Class(Suite))),
+                            trees=Dict(String()),
```
Comment (spbnick) [inline kpet/data.py]: Please don't put `Succession` around the whole structure, put it just around this dictionary's value type. E.g. write something like this:

```python
    arches_schema = Reduction(Regex(), lambda x: [x], List(Regex()))
    # TODO Discard Succession once the database transitioned to the new type
    tree_schema = Succession(
        String(),
        lambda x: dict(template=x),
        Struct(required=dict(template=String()),
               optional=dict(arches=arches_schema))
    )
    # ...
    trees=Dict(tree_schema),
    # ...
```

File: kpet/data.py [hunk 11]
```diff
@@ -387,26 +388,105 @@ def validate_host_type_regex(self):
...
+        """
+
+        trees = {}
+        # Create a dict {tree: arches,} where arches is a list of regexes
+        for tree in self.trees:
+            trees[tree] = []
+            if "arches" not in self.trees[tree]:
+                self.trees[tree]["arches"] = [".*"]
+
+            for arch_regex in self.trees[tree]["arches"]:
```
Comment (spbnick) [inline kpet/data.py]: You can save three lines with this:
```python
    for arch_regex in self.trees.get("arches", [re.compile(".*")]):
```
Maybe put `[re.compile(".*")]` into a variable to avoid compilation on every loop.

File: kpet/data.py [hunk 12]
```diff
@@ -387,26 +388,105 @@ def validate_host_type_regex(self):
...
+                temp_trees = []
+                arch_regex = re.compile(arch_regex)
```
Comment (spbnick) [inline kpet/data.py]: You don't need this line, it's already done by the `Regex` schema.

File: kpet/data.py [hunk 13]
```diff
@@ -387,26 +388,105 @@ def validate_host_type_regex(self):
...
+        """
+        Get a list of available arches for each tree
+        Returns:
+            Raises a schema.Invalid exception when finding an invalid regex
+        """
+
+        trees = {}
+        # Create a dict {tree: arches,} where arches is a list of regexes
+        for tree in self.trees:
+            trees[tree] = []
```
Comment (spbnick) [inline kpet/data.py]: You don't need to keep the other tree's arches for each iteration, so just use a single variable. Also, switch to sets to simplify the code and make it shorter. E.g.:

    tree_arches = set()

File: kpet/data.py [hunk 14]
```diff
@@ -387,26 +388,105 @@ def validate_host_type_regex(self):
...
+            if "arches" not in self.trees[tree]:
+                self.trees[tree]["arches"] = [".*"]
+
+            for arch_regex in self.trees[tree]["arches"]:
+                temp_trees = []
```
Comment (spbnick) [inline kpet/data.py]: `temp_trees` is misguiding. Something like `regex_arhes` would be better. You don't really need this line, though. See below.

File: kpet/data.py [hunk 15]
```diff
@@ -387,26 +388,105 @@ def validate_host_type_regex(self):
...
+                arch_regex = re.compile(arch_regex)
+                for arch in self.arches:
+                    if arch_regex.fullmatch(arch):
+                        temp_trees.append(arch)
```
Comment (spbnick) [inline kpet/data.py]: You can replace four of the above lines with this:
```python
    regex_arches = set(filter(arch_regex.fullmatch, self.arches))
```

File: kpet/data.py [hunk 16]
```diff
@@ -387,26 +388,105 @@ def validate_host_type_regex(self):
...
+                if temp_trees == []:
+                    error = ("One of 'trees: arches:' regexes\n" +
+                             "doesn't match any of the available arches\n" +
+                             "The regex: '{0}'\n" +
+                             "The avaliable arches: {1}")
+                    raise Invalid(error, arch_regex.pattern,
+                                  self.arches)
+                trees[tree] += temp_trees.copy()
```
Comment (spbnick) [inline kpet/data.py]: With sets this would be:

    tree_arches |= regex_arches

File: kpet/data.py [hunk 17]
```diff
@@ -387,26 +388,105 @@ def validate_host_type_regex(self):
...
+
+            trees[tree] = list(dict.fromkeys(trees[tree]))  # remove duplicity
+            self.trees[tree]["arches"] = trees[tree]
```
Comment (spbnick) [inline kpet/data.py]: Switching to sets would let you replace the above two lines with this:

    self.trees[tree]["arches"] = list(tree_arches)

File: tests/assets/db/general/index.yaml [hunk 18]
```diff
@@ -1,6 +1,10 @@
 trees:
-    rhel7: trees/rhel7.xml
-    rhel8: trees/rhel8.xml
+    rhel7:
+        template: trees/rhel7.xml
+        arches: .*
```
Comment (spbnick) [inline tests/assets/db/general/index.yaml]: `arches` are optional, so unless you need this to be explicit for some reason, you can omit them.

File: kpet/cmd_tree.py [hunk 19]
```diff
@@ -41,7 +45,18 @@ def main(args):
     database = data.Base(args.db)
     if args.action == 'list':
         regex = re.compile(args.regex) if args.regex else None
-        for tree in sorted(database.trees.keys()):
+        arch_regex = re.compile(args.arch) if args.arch else None
```
Comment (spbnick) [inline kpet/cmd_tree.py]: You can simplify this line and the code below if you do this instead:

    arch_regex = re.compile(args.arch or ".*")

The `regex` above could use that change too (in a separate commit preceding this one).

File: kpet/cmd_tree.py [hunk 20]
```diff
@@ -41,7 +45,18 @@ def main(args):
...
+            matching_trees = []
+            for tree in database.trees:
+                for arch in database.trees[tree]['arches']:
+                    if arch_regex.fullmatch(arch):
+                        matching_trees.append(tree)
+            matching_trees = list(dict.fromkeys(matching_trees))
+        else:
+            matching_trees = database.trees.keys()
+
+        for tree in sorted(matching_trees):
```
Comment (spbnick) [inline kpet/cmd_tree.py]: With the above changes to regex initialization, you can replace this loop and the code above with something like this:
```python
tree_names=[]
for name, value in database.trees.items():
    if regex.fullmatch(name) and
       filter(arch_regex.fullmatch, value['arches']):
        tree_names.append(name)
for name in sorted(tree_names):
    print(name)
```

All of the above goes for `cmd_arch.py` too, more-or-less.

**Review by spbnick — CHANGES_REQUESTED** [review]
Thank you, Ondrej :)! The logic looks good now, I think, I just have a bunch of comments on style and brevity.

File: kpet/cmd_arch.py [hunk 21]
```diff
@@ -41,9 +46,18 @@ def main(args):
...
+                        in database.trees.items()
+                        if tree_regex.fullmatch(key))
+        # Flattens the list of lists and removes duplicates
+        filtered = {y for x in filtered for y in x}
+
+        for arch in list(filter(regex.fullmatch, filtered)):
+            arch_names.add(arch)
+
+        for name in sorted(arch_names):
+            print(name)
```
Comment (Toaster192) [inline kpet/cmd_arch.py]: These two for loops would probably be better to just replace with
```
for arch in sorted(list(filter(regex.fullmatch, filtered))):
    print(arch)
```
And the "arch_names" variable could be removed entirely

File: tests/test_integration.py [hunk 22]
```diff
@@ -61,7 +61,9 @@
                 arches:
                     - arch
                 trees:
-                    tree: tree.xml
+                    tree:
+                        template: tree.xml
+                        arches: .*
```
Comment (Toaster192) [inline tests/test_integration.py]: I will move these test asset changes to another commit

File: kpet/cmd_run.py [hunk 23]
```diff
@@ -181,6 +181,9 @@ def main_create_baserun(args, database):
         raise Exception("Architecture \"{}\" not found".format(args.arch))
     if args.tree not in database.trees:
         raise Exception("Tree \"{}\" not found".format(args.tree))
+    if args.arch not in database.trees[args.tree]['arches']:
```
Comment (spbnick) [inline kpet/cmd_run.py]: Please rebase this.

File: kpet/data.py [hunk 24]
```diff
@@ -387,20 +388,54 @@ def validate_host_type_regex(self):
                                   host_type_regex.pattern,
                                   host_types)
 
+    def convert_tree_arches(self):
+        """
+        Get a list of available arches for each tree
```
Comment (spbnick) [inline kpet/data.py]: This doesn't return anything, so it shouldn't say "get". This should be something like "Convert each tree's supported architecture specification from a list of regexes to a list of architecture names".

File: kpet/cmd_arch.py [hunk 25]
```diff
@@ -41,9 +46,16 @@ def main(args):
...
-            # filter using regex; if it's not set, print all
-            if regex.fullmatch(arch):
-                print(arch)
+        tree_regex = re.compile(args.tree or ".*")
+
+        filtered = list(value['arches'] for (key, value)
+                        in database.trees.items()
+                        if tree_regex.fullmatch(key))
+        # Flattens the list of lists and removes duplicates
+        filtered = {y for x in filtered for y in x}
```
Comment (spbnick) [inline kpet/cmd_arch.py]: This is a bit too complicated. I also have no idea how that set
comprehension syntax works :)

I think this would be simpler to read:

    tree_arches = set()
    for name, value in database.trees.items():
        if tree_regex.fullmatch(name):
            tree_arches |= set(value['arches'])

    for arch in sorted(filter(regex.fullmatch, tree_arches)):
        print(arch)

You don't need to explicitly convert the output of `filter` to a list as
`sorted` accepts an iterator.

**Review by spbnick — CHANGES_REQUESTED** [review]
Thank you, Ondrej! I think this is almost perfect. Just a couple more style comments.

File: kpet/cmd_run.py (hunk 23 above)
Comment (Toaster192) [reply kpet/cmd_run.py]: Split the schema changes, arch/tree and cmd_run changes

Comment (spbnick) [reply kpet/cmd_run.py]: Please note that this will produce broken code upon automatic rebasing.

**Review by spbnick — CHANGES_REQUESTED** [review]
Everything looks good now, but please rebase on latest master and note the comment inline regarding the automatic rebase.
Thank you, Ondrej :)

**Comment by spbnick** [comment]
@cki-bot, please test this!

**Review by spbnick — APPROVED** [review]

**Comment by spbnick** [comment]
I'm gonna go on a limb here and declare failing pipelines unrelated to this PR. Merging!
